server/src/main/java/org/apache/uniffle/server/ShuffleFlushManager.java

```diff
@@ -143,86 +143,97 @@ public void addToFlushQueue(ShuffleDataFlushEvent event) {
+        Storage storage = storageManager.selectStorage(event);
+        if (storage == null) {
+          break;
+        }
+
+        if (!storage.canWrite()) {
+          // todo: Could we add an interface supportPending for storageManager
+          //       to unify following logic of multiple different storage managers
+          if (storageManager instanceof MultiStorageManager) {
+            event.increaseRetryTimes();
+            ShuffleServerMetrics.incStorageRetryCounter(storage.getStorageHost());
+            continue;
+          } else {
+            // To avoid being re-pushed to pending queue and make the server too much pressure,
+            // it's better to drop directly.
+            if (event.isPended()) {
               break;
-            } else {
-              event.increaseRetryTimes();
-              ShuffleServerMetrics.incStorageRetryCounter(storage.getStorageHost());
             }
-          } while (event.getRetryTimes() <= retryMax);
+            event.increaseRetryTimes();
+            event.markPended();
+            addPendingEvents(event);
+            return;
+          }
         }
-      }
-    } catch (Exception e) {
-      // just log the error, don't throw the exception and stop the flush thread
-      LOG.error("Exception happened when process flush shuffle data for " + event, e);
-    } finally {
-      cleanupFlushEventData(event);
-      if (shuffleServer != null) {
-        long duration = System.currentTimeMillis() - start;
+
+        String user = StringUtils.defaultString(
+            shuffleServer.getShuffleTaskManager().getUserByAppId(event.getAppId()),
+            StringUtils.EMPTY
+        );
+        CreateShuffleWriteHandlerRequest request = new CreateShuffleWriteHandlerRequest(
+            storageType,
+            event.getAppId(),
+            event.getShuffleId(),
+            event.getStartPartition(),
+            event.getEndPartition(),
+            storageBasePaths.toArray(new String[storageBasePaths.size()]),
+            shuffleServerId,
+            hadoopConf,
+            storageDataReplica,
+            user);
+        ShuffleWriteHandler handler = storage.getOrCreateWriteHandler(request);
+        writeSuccess = storageManager.write(storage, handler, event);
         if (writeSuccess) {
-          LOG.debug("Flush to file success in " + duration + " ms and release " + event.getSize() + " bytes");
+          updateCommittedBlockIds(event.getAppId(), event.getShuffleId(), blocks);
+          ShuffleServerMetrics.incStorageSuccessCounter(storage.getStorageHost());
+          break;
         } else {
-          LOG.error("Flush to file for " + event + " failed in "
-              + duration + " ms and release " + event.getSize() + " bytes");
+          event.increaseRetryTimes();
+          ShuffleServerMetrics.incStorageRetryCounter(storage.getStorageHost());
         }
+      } catch (Exception e) {
```

Comment:
Exception -> Throwable ?

Reply:
For what? I can't get this point.

Reply:
For catch more exceptions. Throwable will be more safe.

Reply:
@jerqi Do you mean need to catch `Error`s?

Reply:
> @jerqi Do you mean need to catch `Error`s?

Yes, I want to catch more exceptions and errors because we don't use `finally` to release our resource. If errors will happen, the resource will leak.

Reply:
Done